conbench/entities/run.py

```diff
@@ -123,101 +143,149 @@ def create(data) -> "Run":
-        ).first()
+        # Figure out a list of commits that were skipped in the search for a baseline
+        commit_hashes_searched = Session.scalars(
+            s.select(Commit.sha)
+            .select_from(ancestor_commits)
+            .join(Commit, Commit.id == ancestor_commits.c.ancestor_id)
+            .order_by(ancestor_commits.c.commit_order.desc())
+        ).all()
+        index_of_baseline = commit_hashes_searched.index(baseline_commit_hash)
+        commits_skipped = commit_hashes_searched[:index_of_baseline]
+
+        return _CandidateBaselineSearchResult(
+            baseline_run_id=baseline_run_id, commits_skipped=commits_skipped
+        )
+
+    def get_candidate_baseline_runs(self) -> Dict[str, dict]:
+        """Return information about a few different candidate baseline runs, including
+        on the parent commit, fork-point commit, and head-of-default-branch commit.
 
-        if not closest_run_id:
-            log.debug(
-                "Could not find a matching benchmark_result in this Run's ancestry"
+        See docstring of _search_for_baseline_run() for how these are found.
+        """
+        candidates: Dict[str, _CandidateBaselineSearchResult] = {}
+
+        # The direct, single parent in the git graph
+        if not self.commit:
+            candidates["parent"] = _CandidateBaselineSearchResult(
+                error="the contender run is not connected to the git graph"
+            )
+        else:
+            candidates["parent"] = self._search_for_baseline_run(
+                self.commit.get_parent_commit()
             )
-            return None
 
-        return Run.get(closest_run_id)
+        # If this is a PR run, the PR's fork point commit on the default branch
+        if not self.commit:
+            candidates["fork_point"] = _CandidateBaselineSearchResult(
+                error="the contender run is not connected to the git graph"
+            )
+        elif self.commit.sha == self.commit.fork_point_sha:
+            candidates["fork_point"] = _CandidateBaselineSearchResult(
+                error="the contender run is already on the default branch"
+            )
+        else:
+            candidates["fork_point"] = self._search_for_baseline_run(
+                self.commit.get_fork_point_commit()
+            )
 
-    def get_default_baseline_id(self):
-        run = self.get_default_baseline_run()
-        return run.id if run else None
+        # The latest commit on the default branch that Conbench knows about
+        if self.commit and self.commit.sha == self.commit.fork_point_sha:
+            candidates["latest_default"] = _CandidateBaselineSearchResult(
+                error="the contender run is already on the default branch"
+            )
+        else:
+            # Query the DB for the latest commit
+            query = s.select(Commit).filter(Commit.sha == Commit.fork_point_sha)
+
+            # TODO: how do we filter by repository if there's no commit?
```

Comment:
This relates to @jonkeane 's recent comment: https://github.com/conbench/conbench/pull/1134#issuecomment-1518240762

To support multi-repo Conbenches, we may have to figure out how to submit a run with no commit hash but still a repository, to fully power this `latest_default` candidate baseline run logic. That's not critical right now because it will work as expected for all single-repository Conbench instances (we'll still grab the latest run on the default branch). But if there's a multi-repository Conbench instance, we might grab the latest run on _any_ default branch, which could be confusing.

Reply:
If we're going to leave this as is, please make an issue to come back and clean it up (and we should either prioritize it or clean this up to emit something less seemingly random if someone hits it)

Reply:
Cool, made https://github.com/conbench/conbench/issues/1165